### PR TITLE
Adjust envrionment variables - more sensible defaults

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

The new setup will now set env variables in one place and in one place
only. The old setup used two separate places wich is not DRY and
confusing.

Some default values changed:

1. `PFLOGSUMM_TRIGGER`: logrotate => none
2. `REPORT_SENDER`: mailserver-report@HOSTNAME => mailserver-report@DOMAIN
3. `REPORT_RECIPIENT`: "0" => `POSTMASTER_ADDRESS`
4. `LOGROATE_INTERVAL`: `daily` => `weekly`

One env variable was renamed: `REPORT_INTERVAL` => `LOGROTATE_INTERVAL`

I believe these defaults to be more sensible, especially the REPORT_RECIPIENT
address. The PFLOGSUMM_TRIGGER value was changed to `none` because otherwise
people would start getting daily Postfix log summary reports automatically.
Now, this is opt-in, and reports are sent only when enabled properly.

Some of the variables changed were marked as deprecated. I removed the note,
as the variables now bear some (sane) defaults again for other variables
(i.e.) `REPORT_RECIPIENT` is now default for other recipient addresses.

Although some defaults changed, this is not a breaking change as the
mail server starts exactly as before. Even for REPORT_RECIPIENT="0"
explictly set, other reports are opt-in as well with their defaults,
and as the recipient default changed, this is fine as well.

I would leave it up to vote whether this PR qualifies for `10.5.0` or for `11.0.0`. I'm also open to other defaults, although I think they're okay as they are. Please watch out for inconsistencies in the documentation :D

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: docs/content/config/environment.md [hunk 1]
```diff
@@ -293,49 +293,32 @@ Sender address (`FROM`) for logwatch reports if logwatch reports are enabled.
...
 - monthly => Rotate monthly.
 
-Note that only the log inside the container is affected.
-The full log output is still available via `docker logs mailserver` (_or your respective container name_).
-If you want to control logrotation for the docker generated logfile, see: [Docker Logging Drivers](https://docs.docker.com/config/containers/logging/configure/).
+Note: This Variable actually controls logrotate inside the container and rotates the log depending on this setting. The main log output is still available in its entirety via `docker logs mail` (Or your respective container name). If you want to control logrotation for the Docker-generated logfile see: <https://docs.docker.com/config/containers/logging/configure/>.
 
 Also note that by default the logs are lost when the container is recycled. To keep the logs, mount a volume.
 
-Finally the logrotate interval **may** affect the period for generated reports. That is the case when the reports are triggered by log rotation.
```
Comment (polarathene) [inline docs/content/config/environment.md:337]: In regards to the earlier comment about `LOGROTATE_INTERVAL` behaviour description changing, is this why this line is no longer necessary as well?

File: target/scripts/start-mailserver.sh [hunk 2]
```diff
@@ -41,12 +44,17 @@ VARS[FAIL2BAN_BLOCKTYPE]="${FAIL2BAN_BLOCKTYPE:=drop}"
...
 VARS[LDAP_START_TLS]="${LDAP_START_TLS:=no}"
-VARS[LOGROTATE_INTERVAL]="${LOGROTATE_INTERVAL:=${REPORT_INTERVAL:-daily}}"
+VARS[LOGROTATE_INTERVAL]="${LOGROTATE_INTERVAL:=daily}"
 VARS[LOGWATCH_INTERVAL]="${LOGWATCH_INTERVAL:=none}"
+VARS[LOGWATCH_RECIPIENT]="${LOGWATCH_RECIPIENT:=${REPORT_RECIPIENT}}"
+VARS[LOGWATCH_SENDER]="${LOGWATCH_SENDER:=${REPORT_SENDER}}"
 VARS[MOVE_SPAM_TO_JUNK]="${MOVE_SPAM_TO_JUNK:=1}"
 VARS[NETWORK_INTERFACE]="${NETWORK_INTERFACE:=eth0}"
 VARS[ONE_DIR]="${ONE_DIR:=1}"
 VARS[OVERRIDE_HOSTNAME]="${OVERRIDE_HOSTNAME}"
```
Comment (polarathene) [inline target/scripts/start-mailserver.sh:54]: If this line ever receives a fallback value, it will be ignored by `_obtain_hostname_and_domainname` call that has now been moved prior to the declaration/assignment.

**Review by polarathene — CHANGES_REQUESTED** [review]

File: docs/content/config/environment.md [hunk 3]
```diff
@@ -293,49 +293,32 @@ Sender address (`FROM`) for logwatch reports if logwatch reports are enabled.
...
-Note: This variable used to control logrotate inside the container and sent the pflogsumm report when the logs were rotated.
-It is still supported for backwards compatibility, but the new option LOGROTATE_INTERVAL has been added that only rotates
-the logs.
+- **empty** => `mailserver-report@<YOUR DOMAIN>`
+- => Specify the sender address
 
 ##### LOGROTATE_INTERVAL
 
-Defines the interval in which the mail log is being rotated.
+Changes the interval in which a report is being sent.
```
[the remark this answers is not in the document]
Comment (polarathene) [reply docs/content/config/environment.md:312]: Oh, mentioned in the PR description:

> One env variable was renamed: `REPORT_INTERVAL` => `LOGROTATE_INTERVAL`

[the remark this answers is not in the document]
Comment (georglauterbach) [reply docs/content/config/environment.md:312]: Ah, yes, I think you spotted something there. The `LOGROTATE_INTERVAL` is not responsible for reports being sent. I will need to fix this.

File: docs/content/config/environment.md (hunk 1 above)
Comment (georglauterbach) [reply docs/content/config/environment.md:337]: Regarding the comment above, this is exactly the information that got lost but shouldn't have. I will revive it at an appropriate place.

File: target/scripts/start-mailserver.sh (hunk 2 above)
Comment (georglauterbach) [reply target/scripts/start-mailserver.sh:54]: Which line exactly do you mean? `VARS[OVERRIDE_HOSTNAME]="${OVERRIDE_HOSTNAME}"`? I cannot identify a line that would be negatively impacted. I spotted a missing `:-` though for good style.

File: docs/content/config/environment.md [hunk 4]
```diff
@@ -293,49 +293,32 @@ Sender address (`FROM`) for logwatch reports if logwatch reports are enabled.
...
 
 - **daily** => Rotate daily.
 - weekly => Rotate weekly.
 - monthly => Rotate monthly.
 
-Note that only the log inside the container is affected.
-The full log output is still available via `docker logs mailserver` (_or your respective container name_).
-If you want to control logrotation for the docker generated logfile, see: [Docker Logging Drivers](https://docs.docker.com/config/containers/logging/configure/).
+Note: This Variable actually controls logrotate inside the container and rotates the log depending on this setting. The main log output is still available in its entirety via `docker logs mail` (Or your respective container name). If you want to control logrotation for the Docker-generated logfile see: <https://docs.docker.com/config/containers/logging/configure/>.
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply docs/content/config/environment.md:318]: I think it is fine as is, but I can tend to making it more precise with an upcoming commit / push.

File: target/scripts/start-mailserver.sh (hunk 2 above)
Comment (georglauterbach) [reply target/scripts/start-mailserver.sh:54]: Now I see what you're referring to. I think this is fine as we can check for fallback again in `_obtain_hostname_and_domainname` if the variable is not set.

File: docs/content/config/environment.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (georglauterbach) [reply docs/content/config/environment.md:312]: solved in [afd1942](https://github.com/docker-mailserver/docker-mailserver/pull/2428/commits/afd1942e9511f6efba0d0aaea79bd21e9c550620)

File: docs/content/config/environment.md (hunk 1 above)
Comment (georglauterbach) [reply docs/content/config/environment.md:337]: solved in [afd1942](https://github.com/docker-mailserver/docker-mailserver/pull/2428/commits/afd1942e9511f6efba0d0aaea79bd21e9c550620)

**Review by polarathene — DISMISSED** [review]

**Comment by georglauterbach** [comment]
@polarathene @casperklein @wernerfred do you think this can go into `10.5.0` or should it go into `11.0.0`?

**Comment by wernerfred** [comment]
Refactoring is 10.4.1 but changing the Default is 11.0 from my point of view

**Comment by georglauterbach** [comment]
As much as I'd like to see this in `10.5.0`, I understand that this may not be fitting for the next minor release. I prepare the minor and we need to decide on this and on @casperklein's PR #2424. Afterwards, we can enable the feature freeze, wait one week, release `v10.5.0` and then merge this and #2424.

**Review by casperklein — CHANGES_REQUESTED** [review]

File: mailserver.env [hunk 5]
```diff
@@ -216,24 +216,28 @@ LOGWATCH_RECIPIENT=
...
 #
-# Note: This Variable actually controls logrotate inside the container and rotates the log depending on this setting. The main log output is still available in its entirety via `docker logs mail` (Or your respective container name). If you want to control logrotation for the docker generated logfile see: [Docker Logging Drivers](https://docs.docker.com/config/containers/logging/configure/)
-REPORT_INTERVAL=daily
+# Note: This Variable actually controls logrotate inside the container
+# and rotates the log depending on this setting. The main log output is
+# still available in its entirety via `docker logs mail` (Or your
+# respective container name). If you want to control logrotation for
+# the Docker-generated logfile see:
+# https://docs.docker.com/config/containers/logging/configure/
+LOGROTATE_INTERVAL=daily
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply mailserver.env:240]: Ah yes, I changed this at first but reverted back later but forgot to update the docs. Thanks!

File: mailserver.env [hunk 6]
```diff
@@ -260,15 +264,7 @@ ENABLE_SPAMASSASSIN=0
 # deliver spam messages in the inbox (eventually tagged using SA_SPAM_SUBJECT)
 SPAMASSASSIN_SPAM_TO_INBOX=1
 
-# KAM is a 3rd party SpamAssassin ruleset, provided by the McGrail Foundation.
-# If SpamAssassin is enabled, KAM can be used in addition to the default ruleset.
-# - **0** => KAM disabled
-# - 1 => KAM enabled
-#
-# Note: only has an effect if `ENABLE_SPAMASSASSIN=1`
-ENABLE_SPAMASSASSIN_KAM=0
```
Comment (casperklein) [inline mailserver.env:269]: Do I miss something or why was this removed?

Comment (georglauterbach) [reply mailserver.env:269]: You missed nothing, I'm an id... this is a nice copy-waste-error... I'm correcting this.

Comment (georglauterbach) [reply mailserver.env:269]: correction done in c026d27130c95e787eba0acc4af7e28162bd32b0

File: target/scripts/start-mailserver.sh [hunk 7]
```diff
@@ -3,19 +3,22 @@
...
 
+_obtain_hostname_and_domainname
+
 # ------------------------------------------------------------
 # ? <<
 # --
 # ? >> Setup of default and global values / variables
 # ------------------------------------------------------------
 
+VARS[POSTMASTER_ADDRESS]="${POSTMASTER_ADDRESS:=postmaster@${DOMAINNAME}}"
```
Comment (casperklein) [inline target/scripts/start-mailserver.sh:18]:
```suggestion
# These variables must be defined first; They are used as default values for other variables.
VARS[POSTMASTER_ADDRESS]="${POSTMASTER_ADDRESS:=postmaster@${DOMAINNAME}}"
```

Comment (georglauterbach) [reply target/scripts/start-mailserver.sh:18]: I honestly believe this to be very obvious, but if other see it as @casperklein, I will add this without further objection.

Comment (casperklein) [reply target/scripts/start-mailserver.sh:18]: I needed some seconds to realize, why there was a separation to the other variables 😆 I was afraid, that someone else in the future tries to "optimize" it, by removing the empty line and sorting the variables alphabetically.

Comment (polarathene) [reply target/scripts/start-mailserver.sh:18]: I don't mind either way, I assume the comment was about some of the `VARS[ ... ]` values being defined earlier and used by others below.

Maybe there is value in making that more clear upfront to any new contributors / maintainers that are less familiar, but at the same time, I'd expect breakage or failing tests to catch such dependencies being defined later than they should, or at the very least in review from the core maintainers we have.

Comment (polarathene) [reply target/scripts/start-mailserver.sh:18]: > I needed some seconds to realize, why there was a separation to the other variables

I had the same reaction initially when first reviewing, but I think that's a one off time.

I don't know if the comment would help much if someone decides to collapse the space separation, would probably be helpful in that case to have a more clear separation (eg: `# ==========` or similar), but that isn't an area that should be getting worked on that often that we'd risk missing such a mistake happening in review?

Comment (georglauterbach) [reply target/scripts/start-mailserver.sh:18]: Alright, I will add the comment, but I don't think we need more / extra separators :D

File: target/scripts/start-mailserver.sh [hunk 8]
```diff
@@ -41,12 +45,17 @@ VARS[FAIL2BAN_BLOCKTYPE]="${FAIL2BAN_BLOCKTYPE:=drop}"
 VARS[FETCHMAIL_PARALLEL]="${FETCHMAIL_PARALLEL:=0}"
 VARS[FETCHMAIL_POLL]="${FETCHMAIL_POLL:=300}"
 VARS[LDAP_START_TLS]="${LDAP_START_TLS:=no}"
-VARS[LOGROTATE_INTERVAL]="${LOGROTATE_INTERVAL:=${REPORT_INTERVAL:-daily}}"
+VARS[LOGROTATE_INTERVAL]="${LOGROTATE_INTERVAL:=daily}"
```
Comment (casperklein) [inline target/scripts/start-mailserver.sh:48]:
```suggestion
VARS[LOGROTATE_INTERVAL]="${LOGROTATE_INTERVAL:=weekly}"
```

Comment (georglauterbach) [reply target/scripts/start-mailserver.sh:48]: Now I see committing the suggestion in https://github.com/docker-mailserver/docker-mailserver/pull/2428#discussion_r812365051 was an error on my side. We can go ahead and change the default of `LOGROTATE_INTERVAL` to weekly, but then we'd need to do it uniformly. The old default value was `daily`. If you agree @casperklein, I will change it to weekly (then I'd also need to change the docs and some tests). Otherwise, if we want to stay at the current default, it'd need to revert https://github.com/docker-mailserver/docker-mailserver/pull/2428#discussion_r812365051 for consistency.

File: mailserver.env (hunk 5 above)
[the remark this answers is not in the document]
Comment (georglauterbach) [reply mailserver.env]: Commmiting this may have been an error on my side. The default actually is `daily`, not `weekly`.

File: target/scripts/start-mailserver.sh (hunk 8 above)
Comment (casperklein) [reply target/scripts/start-mailserver.sh:48]: The logs are rotated 4 times. I think `weekly` is a better default, otherwise there isn't a complete week of logs. This is the minimum, one should have in case of errors for debugging.

Comment (georglauterbach) [reply target/scripts/start-mailserver.sh:48]: Then I will change it consistently to `weekly` with the next commit.

Comment (georglauterbach) [reply target/scripts/start-mailserver.sh]: resolved in [ad7e1f7](https://github.com/docker-mailserver/docker-mailserver/pull/2428/commits/ad7e1f72bafdad7f5dbbb94d05f23cfb06ffc93d)

File: mailserver.env (hunk 5 above)
[the remark this answers is not in the document]
Comment (georglauterbach) [reply mailserver.env]: [ad7e1f7](https://github.com/docker-mailserver/docker-mailserver/pull/2428/commits/ad7e1f72bafdad7f5dbbb94d05f23cfb06ffc93d)

**Review by polarathene — APPROVED** [review]

**Review by casperklein — DISMISSED** [review]

**Review by casperklein — APPROVED** [review]

**Review by polarathene — APPROVED** [review]
